## Reading Data

Relevant imports go here:

In [11]:
import pandas as pd
import numpy as np
from IPython.display import display

Read in the different csv files

In [2]:
import pandas as pd

routes_df = pd.read_csv('data_output/routes.csv')
countries_df = pd.read_csv('data_output/countries.csv')
planes_df = pd.read_csv('data_output/planes.csv')
airports_df = pd.read_csv('data_output/airports.csv', on_bad_lines='skip')
airlines_df = pd.read_csv('data_output/airlines.csv', on_bad_lines='skip')

Inspect briefly the dataframes

In [3]:
# display(routes_df.head())
# display(planes_df.head())
# display(countries_df.head())
# display(airports_df.head())
# display(airlines_df.head())

## Data Cleaning 

### Cleaning for All Dataframes

Upon inspecting the dataframes, we find a few fixes that need to be made before proceeding into the data analysis stage. These issues + their relevant fixes will be highlighted in this section.

All the dataframes have their values in the cells within the quotation marks. Here, we will get rid of the quotation marks using the replace function with regular expressions. So, for instance, the name of an airplane will go from **"Aerospatiale (Nord) 262"** to **Aerospatiale (Nord) 262**

In [4]:
routes_df = routes_df.replace('"', '', regex=True)
planes_df = planes_df.replace('"', '', regex=True)
countries_df = countries_df.replace('"', '', regex=True)
airports_df = airports_df.replace('"', '', regex=True)
airlines_df = airlines_df.replace('"', '', regex=True)

### Cleaning for routes_df

We start by performing a general inspection of the routes_df dataframe.

In [138]:
display(routes_df.describe(include='all'))
display(routes_df.info())

Airline Airline ID Source Airport Source Airport ID  \
count    67663      67663          67663             67663   
unique     568        548           3409              3321   
top         FR       4296            ATL              3682   
freq      2484       2484            915               915   
mean       NaN        NaN            NaN               NaN   
std        NaN        NaN            NaN               NaN   
min        NaN        NaN            NaN               NaN   
25%        NaN        NaN            NaN               NaN   
50%        NaN        NaN            NaN               NaN   
75%        NaN        NaN            NaN               NaN   
max        NaN        NaN            NaN               NaN   

       Destination Airport Destination Airport ID Codeshare         Stops  \
count                67663                  67663     14597  67663.000000   
unique                3418                   3327         1           NaN   
top                    ATL                   3682         Y           NaN   
freq                   911                    911     14597           NaN   
mean                   NaN                    NaN       NaN      0.000163   
std                    NaN                    NaN       NaN      0.012749   
min                    NaN                    NaN       NaN      0.000000   
25%                    NaN                    NaN       NaN      0.000000   
50%                    NaN                    NaN       NaN      0.000000   
75%                    NaN                    NaN       NaN      0.000000   
max                    NaN                    NaN       NaN      1.000000   

       Equipment  
count      67645  
unique      3925  
top          320  
freq        9180  
mean         NaN  
std          NaN  
min          NaN  
25%          NaN  
50%          NaN  
75%          NaN  
max          NaN

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 67663 entries, 0 to 67662
Data columns (total 9 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   Airline                 67663 non-null  object
 1   Airline ID              67663 non-null  object
 2   Source Airport          67663 non-null  object
 3   Source Airport ID       67663 non-null  object
 4   Destination Airport     67663 non-null  object
 5   Destination Airport ID  67663 non-null  object
 6   Codeshare               14597 non-null  object
 7   Stops                   67663 non-null  int64 
 8   Equipment               67645 non-null  object
dtypes: int64(1), object(8)
memory usage: 4.6+ MB


None

Of the first items of interest, we find that the codeshare column is mainly a column that is filled with 'NaN' values. Further inspection upon the usefulness of the column indicates low relevance for the purpose of this analysis. The codeshare column will appear as "Y" if the particular flight is not operated by the airline but by another carrier. Therefore, we will first drop this column. 

In [139]:
routes_df.drop(columns=['Codeshare'], inplace=True)

Next, we note that there are some rows in the equipment column that have null objects. Because the equipment column (which shows what types of aircrafts are mainly used for the routes) is an important one, we will omit all the rows that have null equipment values.

In [140]:
routes_df = routes_df.dropna(subset=['Equipment'])

Continuing the work on the 'equipment' column, we find that in some cases, the column has multiple equipment all listed in the same cell. For instance, take rows 64049 to 64055 as shown below.

In [141]:
routes_df.iloc[64049:64056]

Airline Airline ID Source Airport Source Airport ID Destination Airport  \
64067      WN       4547            MCO              3878                 PHX   
64068      WN       4547            MCO              3878                 PIT   
64069      WN       4547            MCO              3878                 PVD   
64070      WN       4547            MCO              3878                 RDU   
64071      WN       4547            MCO              3878                 RIC   
64072      WN       4547            MCO              3878                 ROC   
64073      WN       4547            MCO              3878                 SAT   

      Destination Airport ID  Stops        Equipment  
64067                   3462      0          73W 738  
64068                   3570      0              73W  
64069                   3641      0  73C 733 73W 73H  
64070                   3626      0      73W 733 73C  
64071                   3608      0      733 73W 73C  
64072                   3622      0              73W  
64073                   3621      0          73H 73W

As shown, some of the cells in the 'Equipment' column have more than one equipment/aircraft type listed within the cell. We would like to have a separate row for each of the equipment in order for appropriate analysis. Therefore, we need to separate a row which has the cell "73C 733 73W 73H" into four different rows so that every type of equipment has its own row. The rows would be: 
- Row 1 would have the 73C in the 'Equipment' column
- Row 2 would have the 733 in the 'Equipment' column
- Row 3 would have the 73W in the 'Equipment' column
- Row 4 would have the 73H in the 'Equipment' column

Therefore, the last step in cleaning the routes_df would be to 'explode' these rows with multiple strings in the 'Equipment' column into different rows for each equipment. 

In [142]:
routes_df['Equipment'] = routes_df['Equipment'].str.split()
routes_df = routes_df.explode('Equipment')

### Cleaning for countries_df

We start by performing a general inspection of the countries_df dataframe.

In [119]:
display(countries_df.describe(include='all'))
display(countries_df.info())

name iso_code dafif_code Unnamed: 3
count     261      261        261          2
unique    259      241        260          2
top     India       \N         BQ           
freq        2       19          2          1

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 261 entries, 0 to 260
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   name        261 non-null    object
 1   iso_code    261 non-null    object
 2   dafif_code  261 non-null    object
 3   Unnamed: 3  2 non-null      object
dtypes: object(4)
memory usage: 8.3+ KB


None

Whilst inspecting the countries_df, we found a few items of interest. First, we find that there are 261 rows in the countries dataframe, but only 259 are unique (as seen in the unique column). Let's inspect further. 

In [120]:
duplicate_countries = countries_df[countries_df.duplicated(subset='name', keep=False)]
print(duplicate_countries)

          name iso_code dafif_code Unnamed: 3
33       India       IN         BS        NaN
101  Palestine       PS         GZ        NaN
112      India       IN         IN        NaN
252  Palestine       PS         WE        NaN


The reason for the duplicates is because the country 'India' and 'Palestine; has two different dafif_codes. Further research shows that the "DAFIF code" is not a standard or widely recognized code for countries. It's possible that "DAFIF" refers to a specific system or dataset that uses its own set of codes for countries or locations. DAFIF (Digital Aeronautical Flight Information File) codes are mainly used in the aviation and aeronautical navigation field and the information is not readily available. Seeing that the dafif_code is not as popular, and that the ISO code is more standard, this will guide us to focus on the ISO_codes from now. For the purpose of cleaning and avoiding duplicates, we will get rid of the first of the duplicates for each of the duplicated countries.

The reason for the duplicates is because the country 'India' has two different dafif_codes. The correct dafif_code for India is "IN". Therefore, we will drop the first of the two rows in india_rows (index = 33). 

In [121]:
countries_df = countries_df.drop_duplicates(subset=['name'], keep='last')

name iso_code dafif_code Unnamed: 3
0                 Bonaire       BQ         BQ           
1                   Aruba       AW         AA        NaN
2     Antigua and Barbuda       AG         AC        NaN
3    United Arab Emirates       AE         AE        NaN
4             Afghanistan       AF         AF        NaN
..                    ...      ...        ...        ...
256                 Samoa       WS         WS        NaN
257              Eswatini       SZ         WZ        NaN
258                 Yemen       YE         YM        NaN
259                Zambia       ZM         ZA        NaN
260              Zimbabwe       ZW         ZI        NaN

[259 rows x 4 columns]

We also note that the discrepancy in ISO_codes where, there are 241 unique ISO codes and 259 (now) unique countries. Referring back to the guide from the website that provides the dataset, we find the following "Some entries have DAFIF codes, but not ISO codes. These are primarily uninhabited islands without airports, and can be ignored for most purposes." Given that these locations do not have airports, they become useless for the purpose of our research. Therefore, we will omit all the rows where iso_code == '\N'.

In [115]:
countries_df = countries_df[countries_df['iso_code'] != '\\N']

name                   iso_code dafif_code Unnamed: 3
0                 Bonaire   Saint Eustatius and Saba         BQ           
1                   Aruba                         AW         AA        NaN
2     Antigua and Barbuda                         AG         AC        NaN
3    United Arab Emirates                         AE         AE        NaN
4             Afghanistan                         AF         AF        NaN
..                    ...                        ...        ...        ...
256                 Samoa                         WS         WS        NaN
257              Eswatini                         SZ         WZ        NaN
258                 Yemen                         YE         YM        NaN
259                Zambia                         ZM         ZA        NaN
260              Zimbabwe                         ZW         ZI        NaN

[240 rows x 4 columns]

Now that we have cleaned up the countries_df, we go back to obsolete nature of the dafif_code column and the "Unnamed: 3" column. The final step in making this dataframe analysis-ready is to drop the irrelevant columns. We drop the last two columns of the dataframe. 

Some entries have DAFIF codes, but not ISO codes. These are primarily uninhabited islands without airports, and can be ignored for most purposes.

In [122]:
countries_df = countries_df.iloc[:, :-2]

name iso_code
0                 Bonaire       BQ
1                   Aruba       AW
2     Antigua and Barbuda       AG
3    United Arab Emirates       AE
4             Afghanistan       AF
..                    ...      ...
256                 Samoa       WS
257              Eswatini       SZ
258                 Yemen       YE
259                Zambia       ZM
260              Zimbabwe       ZW

[259 rows x 2 columns]

In [ ]:
# display(planes_df.describe(include='all'))
# display(planes_df.info())

# display(airports_df.describe(include='all'))
# display(airports_df.info())

# display(airlines_df.describe(include='all'))
# display(airlines_df.info())

## Cleaning for planes_df

Before delving into the cleaning for the planes_df, we would like to clarify the following:

- The original planes_df only had the 'Name', 'IATA Code' and 'ICAO code' columns. Upon a visual inspection of the csv file, we noticed that for some of the fields (mainly the smaller/uncommon planes), the IATA or ICAO code was missing. Whilst the number of missing IATA/ICAO codes missing were not significant enough for us to be worried about the completeness data, we handled the missing information by looking the missing data points and filling them in manually. This would be the most direct and time-efficient solution to employ.
- The 'General Make' column was also manually created mainly to accommodate for the Boeings and Airbuses. Acknowledging the huge duopoly that Boeing and Airbus have in terms of airplanes, we are cognizant that a significant number of comparisons and flights being analysed will pertain to aircrafts manufactured by either of the companies. By creating the 'General Make' column, we would like to appropriately account for the many variations of a single model of an aircraft that both companies create.
   - For instance, the Boeing 737 comes in several variations: Boeing 737 MAX 10, Boeing 737 MAX 7, Boeing 737 MAX 8, Boeing 737 MAX 9, Boeing 737-200, Boeing 737-300, Boeing 737-400, Boeing 737-500, etc. Whilst the variations provide a significant amount of detail, we recognize that it may help to group them all as "B737". Similar can be said for the Airbus340 which comes in 5 different makes therefore, the ICAO of the Airbus340 is "A340" which is what will be used to fill in all of the rows that have an Airbus340 make. The goal of the "General Make" is to avoid dwelling on the specifics of the model and rather recognize the general model. For other aircraft makes of aircrafts (for instance Lockheed Martin, Aerospatiale, etc), if the makes shared various types of the model, we also labelled them similarly (based on the main model). If an aircraft is the only single make in the model type, we use the details from the "ICAO code" column and copy it to the "General Make" column.
- For some airplane types, especially the really small/uncommon airplanes, we were unable to find the right IATA codes. Therefore, we have left those rows as is. It is very unlikely that airplanes that do not have an IATA code will be utilized as transportation vehicles on many flight routes anyway.


With this, let us take a look at the planes_df briefly.

In [5]:
planes_df.head()

Name IATA Code ICAO Code  \
0                       Aerospatiale (Nord) 262       ND2      N262   
1  Aerospatiale (Sud Aviation) Se.210 Caravelle       CRV      S210   
2                  Aerospatiale SN.601 Corvette       NDC      S601   
3                Aerospatiale/Alenia ATR 42-300       AT4      AT43   
4                Aerospatiale/Alenia ATR 42-500       AT5      AT45   

  General Make  
0         N262  
1         S210  
2         S601  
3         AT43  
4         AT45

In [6]:
display(planes_df.describe(include='all'))
display(planes_df.info())

Name IATA Code ICAO Code General Make
count                       246       246       246          246
unique                      246       223       243          186
top     Aerospatiale (Nord) 262        \N      B737         B737
freq                          1         9         3           13

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 246 entries, 0 to 245
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Name          246 non-null    object
 1   IATA Code     246 non-null    object
 2   ICAO Code     246 non-null    object
 3   General Make  246 non-null    object
dtypes: object(4)
memory usage: 7.8+ KB


None

From the tables, the item that grabs our attention immediately is the number of unique IATA codes. There are some empty cells for the IATA code of smaller/uncommon aircrafts because they do not have an IATA code. To handle these, we can change the '\N' cells in the IATA column into cells with null values. 

In [12]:
planes_df['IATA Code'] = planes_df['IATA Code'].replace(r'\N', np.nan)
display(planes_df.describe(include='all'))
display(planes_df.info())

Name IATA Code ICAO Code General Make Company
count                       246       237       246          246     246
unique                      246       222       243          186      39
top     Aerospatiale (Nord) 262       CNJ      B737         B737  Boeing
freq                          1         8         3           13      47

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 246 entries, 0 to 245
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Name          246 non-null    object
 1   IATA Code     237 non-null    object
 2   ICAO Code     246 non-null    object
 3   General Make  246 non-null    object
 4   Company       246 non-null    object
dtypes: object(5)
memory usage: 9.7+ KB


None

We notice that the names for the airplanes are really long. Whilst this is good because it gives us some detail, we would also like to do a comparison of the Boeings vs Airbus vs other airplanes. For this, we will create a new column that takes the first string of the "Name" column which is most often the name of the company. 

In [13]:
planes_df['Company'] = planes_df['Name'].str.split().str[0]

To make the analysis more analysis-ready, we can add a new column acknowledging that most of the airplanes for routes will be Boeings and Airbuses. We can make a new column called "airbus_boeing_other" which allocates all Airbus airplanes from the "Company" column as "Airbus" and labels all the Boeing aircrafts as "Boeing". However, for all of the other aircrafts that are not Boeings and Airbuses, these would be labelled as "Other".

In [20]:
planes_df['airbus_boeing_other'] = np.where(planes_df['Company'].isin(['Boeing', 'Airbus']), planes_df['Company'], 'Other')
display(planes_df['airbus_boeing_other'].value_counts())

airbus_boeing_other
Other     172
Boeing     47
Airbus     27
Name: count, dtype: int64

planes_df is now analysis ready! 

## Cleaning for airports_df

In [ ]:
# display(airports_df.describe(include='all'))
# display(airports_df.info())

# display(airlines_df.describe(include='all'))
# display(airlines_df.info())